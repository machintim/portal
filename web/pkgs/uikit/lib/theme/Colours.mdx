import { Meta, Title, Subtitle, ColorPalette, ColorItem } from '@storybook/addon-docs/blocks';
import theme from './defaultTheme';

<Meta title="theme/Colours" />
<Title>Colours explorer</Title>
<Subtitle>Colours approved for use in the UCL Design System</Subtitle> 

More information on the use of colour in the UCL Design System <a href="https://www.figma.com/design/8Sm5PxWOWJYpYXAzhRzUzt/UCL-Design-System-UI-Kit?node-id=1-287" target="_blank">can be found on Figma</a>. 

The colours populating this page are imported from the <a href="https://github.com/ucl-isd/uikit-react/blob/main/lib/theme/defaultTheme.ts#L5" target="_blank">theme file</a>. 

Below are all colours per category, one category per line: 

<ColorPalette>
{
  Object.entries(theme.color).map(([key, value]) => {
    return (
      <ColorItem key={key} title={key.charAt(0).toUpperCase() + key.slice(1)} colors={value} />
    );
  })
}
</ColorPalette>

Below is a breakdown of each colour category. These are the same colours as above, but grouped by category, one colour per line: 

{Object.entries(theme.color).map(([key, value]) => {
  return (
    <div key={key}>
      <h2>{key.charAt(0).toUpperCase() + key.slice(1)}</h2>

      <ColorPalette>

      {Object.entries(value).map(([colorKey, colorValue]) => {
        return (
          <ColorItem key={colorKey} title={colorKey} colors={{ [colorKey]: colorValue }} />
        );
      })}

      {console.log(key, value)}

      </ColorPalette>
    </div>
  );
})}
