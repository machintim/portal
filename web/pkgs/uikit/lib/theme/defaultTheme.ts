// UCL Theme

// base colours

const purple200 = '#2C0442'; // UCL dark purple
const purple50 = '#500778'; // UCL vibrant purple
const purple40 = '#6345A5';
const purple20 = '#BBAACC';
// TODO: where is UCL muted purple = #C6B0BC ?
const green100 = '#006B0F';
const green70 = '#008613';
const green40 = '#80C289';
const green10 = '#E6F3E8';
const green200 = '#113B3A'; // UCL dark green
const green50 = '#52C152'; // UCL vibrant green
const green20 = '#C9D1A8'; // UCL muted green

const blue100 = '#004089';
const blue80 = '#004EA8';
const blue70 = '#0D68CF';
const blue40 = '#7FB7F7';
const blue10 = '#E5F0FD';
const blue5 = '#F2F8FE';
const blue200 = '#002248'; // UCL dark blue
const blue50 = '#34C6C6'; // UCL vibrant blue
const blue20 = '#B6DCE5'; // UCL muted blue

const yellow200 = '#504D48'; // UCL dark yellow
const yellow50 = '#FFCA36'; // UCL vibrant yellow
const yellow20 = '#DAD6CA'; // UCL muted yellow

const orange100 = '#CC7100';
const orange70 = '#FF8D00';
const orange40 = '#FFC680';
const orange10 = '#FFF4E6';

const red100 = '#811712';
const red70 = '#CA0007';
const red50 = '#E48083';
const red40 = '#E48083';
const red10 = '#FAE6E7';

const pink200 = '#4B0A32'; // UCL dark pink
const pink50 = '#AC145A'; // UCL vibrant pink
const pink20 = '#DEB8C3'; // UCL vibrant pink

const black100 = '#000000';
const black90 = '#1A1A1A';
const black80 = '#333333';
const black60 = '#666666';
const black40 = '#999999';
const black20 = '#CCCCCC';
const black15 = '#D9D9D9';
const black10 = '#E6E6E6';
const black5 = '#F2F2F2';

const white = '#ffffff';

const baseColour = {
  purple200,
  purple50,
  purple40,
  purple20,

  green100,
  green70,
  green40,
  green10,
  green200,
  green50,
  green20,

  blue100,
  blue80,
  blue70,
  blue40,
  blue10,
  blue5,
  blue200,
  blue50,
  blue20,

  yellow200,
  yellow50,
  yellow20,

  orange100,
  orange70,
  orange40,
  orange10,
  // missing orange50?
  red100,
  red70,
  red50,
  red40,
  red10,

  pink200,
  pink50,
  pink20,

  black100,
  black90,
  black80,
  black60,
  black40,
  black20,
  black15,
  black10,
  black5,

  white,
};

const textColours = {
  primary: baseColour.black90, // Primary text, such as body copy, headers, displays, labels.
  secondary: baseColour.black60, // Secondary text, such as stand-first, paragraph summary, breadcrumbs.
  muted: baseColour.black40, // Tertiary text, such as input fields placeholder.
  inverted: baseColour.white, // Text on bold background, for example primary buttons label.

  interactive: baseColour.blue70, // Link, secondary button.
  default: baseColour.black60, // Status default for component label: tabs, toolbar.
  disabledOnBg: baseColour.black40, // Status disabled on background.
  disabled: baseColour.black20, // Status disabled.

  active: baseColour.blue70, // Status active for components' label: tabs, toolbar.
  hover: baseColour.blue100, // Status hover for links and components' label.
  visited: baseColour.purple40, // Status visited for links.

  error: baseColour.red70, // Label for buttons or link with destructive purpose.
  errorHover: baseColour.red100, // Error hover.
  errorDisabled: baseColour.red50, // Error disabled.
};

const interactionColours = {
  blue100: baseColour.blue100,
  blue80: baseColour.blue80,
  blue70: baseColour.blue70,
  blue40: baseColour.blue40,
  blue10: baseColour.blue10,
  blue5: baseColour.blue5,
};

const displayColours = {
  purpleDark: baseColour.purple200,
  purpleVibrant: baseColour.purple50,
  purpleMuted: baseColour.purple20,

  greenDark: baseColour.green200,
  greenVibrant: baseColour.green50,
  greenMuted: baseColour.green20,

  yellowDark: baseColour.yellow200,
  yellowVibrant: baseColour.yellow50,
  yellowMuted: baseColour.yellow20,

  pinkDark: baseColour.pink200,
  pinkVibrant: baseColour.pink50,
  pinkMuted: baseColour.pink20,

  blueDark: baseColour.blue200,
  blueVibrant: baseColour.blue50,
  blueMuted: baseColour.blue20,
};

const systemColours = {
  red100: baseColour.red100,
  red70: baseColour.red70,
  red40: baseColour.red40,
  red10: baseColour.red10,

  orange100: baseColour.orange100,
  orange70: baseColour.orange70,
  orange40: baseColour.orange40,
  orange10: baseColour.orange10,

  green100: baseColour.green100,
  green70: baseColour.green70,
  green40: baseColour.green40,
  green10: baseColour.green10,

  blue100: baseColour.blue100,
  blue80: baseColour.blue80,
  blue70: baseColour.blue70,
  blue40: baseColour.blue40,
  blue10: baseColour.blue10,
  blue5: baseColour.blue5,
};

const linkColours = {
  default: baseColour.blue70,
  hover: baseColour.blue100,
  visited: baseColour.purple40,
  disabled: baseColour.black20,
};

const blackAndWhite = {
  black: baseColour.black100,
  white: baseColour.white,
};

const neutralColours = {
  black: blackAndWhite.black,
  grey90: baseColour.black90,
  grey80: baseColour.black80,
  grey60: baseColour.black60,
  grey40: baseColour.black40,
  grey20: baseColour.black20,
  grey15: baseColour.black15,
  grey10: baseColour.black10,
  grey5: baseColour.black5,
  white: blackAndWhite.white,
};

const overlayColours = {
  blanket: 'rgba(0, 0, 0, 0.5)',
};

const color = {
  text: textColours,
  interaction: interactionColours,
  display: displayColours,
  system: systemColours,
  link: linkColours,
  neutral: neutralColours,
  overlay: overlayColours,
};

// These colour definitions will need to be refactored to match the design system.

const baseSizes = {
  s1: '1px',
  s2: '2px',
  s4: '4px',
  s6: '6px',
  s8: '8px',
  s12: '12px',
  s16: '16px',
  s20: '20px',
  s24: '24px',
  s28: '28px',
  s32: '32px',
  s36: '36px',
  s40: '40px',
  s44: '44px',
  s48: '48px',
  s52: '52px',
  s56: '56px',
  s60: '60px',
  s64: '64px',
  s80: '80px',
};

const margin = {
  m2: baseSizes.s2,
  m4: baseSizes.s4,
  m6: baseSizes.s6,
  m8: baseSizes.s8,
  m12: baseSizes.s12,
  m16: baseSizes.s16,
  m20: baseSizes.s20,
  m24: baseSizes.s24,
  m32: baseSizes.s32,
  m40: baseSizes.s40,
  m48: baseSizes.s48,
  m56: baseSizes.s56,
  m64: baseSizes.s64,
};

const padding = {
  p2: baseSizes.s2,
  p4: baseSizes.s4,
  p6: baseSizes.s6,
  p8: baseSizes.s8,
  p12: baseSizes.s12,
  p16: baseSizes.s16,
  p20: baseSizes.s20,
  p24: baseSizes.s24,
  p32: baseSizes.s32,
  p40: baseSizes.s40,
  p48: baseSizes.s48,
  p56: baseSizes.s56,
  p64: baseSizes.s64,
};

const breakpoints = {
  mobile: 320,
  tablet: 672,
  desktop: 1056,
  desktopLarge: 1536,
};

const createMediaQuery = (breakpoint: number) =>
  // Breakpoints apply at the specified width and above
  `@media (min-width: ${breakpoint}px)`;

const custom = (breakpoint: number) =>
  // Fallback for edge cases
  createMediaQuery(breakpoint);

const mq = {
  mobile: createMediaQuery(breakpoints.mobile),
  tablet: createMediaQuery(breakpoints.tablet),
  desktop: createMediaQuery(breakpoints.desktop),
  desktopLarge: createMediaQuery(breakpoints.desktopLarge),
  custom,
};

const theme = {
  color,

  thickness: {
    t1: baseSizes.s1,
    t2: baseSizes.s2,
  },

  width: {
    h16: baseSizes.s16,
    h20: baseSizes.s20,
    h24: baseSizes.s24,
  },

  height: {
    h16: baseSizes.s16,
    h20: baseSizes.s20,
    h24: baseSizes.s24,
    h32: baseSizes.s32,
    h36: baseSizes.s36,
    h40: baseSizes.s40,
    h80: baseSizes.s80,
  },

  margin,

  padding,

  radius: {
    r2: baseSizes.s2,
    r4: baseSizes.s4,
    r8: baseSizes.s8,
    r12: baseSizes.s12,
    r16: baseSizes.s16,
  },

  border: {
    b1: baseSizes.s1,
    b2: baseSizes.s2,
  },

  outline: {
    o2: baseSizes.s2,
    o4: baseSizes.s4,
  },

  font: {
    family: {
      primary: 'Helvetica, "Helvetica Neue", Arial, sans-serif',
      secondary:
        'Garamond, Baskerville, "Baskerville Old Face", "Hoefler Text", "Times New Roman", serif',
      fixed: 'Courier, fixed',
    },
    size: {
      f12: '12px',
      f14: '14px', // Font should be at least 14px in most cases
      f16: '16px', // Prefer 16px to 14px in most cases
      f18: '18px',
      f20: '20px',
      f22: '22px',
      f24: '24px',
      f26: '26px',
      f28: '28px',
      f30: '30px',
      f32: '32px',
      f36: '36px',
      f40: '40px',
    },
    weight: {
      light: 300,
      regular: 400,
      medium: 500,
      semiBold: 600,
      bold: 700,
      extraBold: 800,
    },
    lineHeight: {
      h100: '100%',
      h110: '110%',
      h120: '120%',
      h130: '130%',
      h140: '140%',
      h150: '150%',
    },
  },

  boxShadow: {
    // offset-x | offset-y | blur-radius | spread-radius | color
    y1: '0 1px 4px 1px rgba(0, 0, 0, 0.1)',
    y2: '0 2px 4px 1px rgba(0, 0, 0, 0.1)',
    x2y4: '2px 4px 8px 4px rgba(0, 0, 0, 0.1)',
    focus: `0px 0px 0px 2px ${color.neutral.white}, 0px 0px 0px 4px ${color.neutral.black}, 0px 0px 0px 6px ${color.system.orange70}`,
    active: `0px 6px 0px 0px ${color.interaction.blue80} inset`,
    destructiveActive: `0px 6px 0px 0px ${color.system.red100} inset`,
  },

  breakpoints,
  mq,

  // For ease of use, margin and padding are added at the top level
  // Usage: margin-top: ${theme.m16};
  ...margin,
  ...padding,
};

export type ThemeType = typeof theme;
export type DisplayColoursType = typeof displayColours;
export type BlackAndWhiteColoursType = typeof blackAndWhite;

export default theme;
