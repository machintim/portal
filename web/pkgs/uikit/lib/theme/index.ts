export { default as theme } from './defaultTheme';
export type {
  ThemeType,
  DisplayColoursType,
  BlackAndWhiteColoursType,
} from './defaultTheme';
export {
  default as useTheme,
  ThemeContextProvider,
} from './useTheme';
