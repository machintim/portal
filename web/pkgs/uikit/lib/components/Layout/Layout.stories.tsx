import type { Meta, StoryObj } from '@storybook/react';
import { css } from '@emotion/css';
import { useTheme } from '../../theme';
import Layout from './Layout';
import { Heading, Button, Text } from '../';

const meta = {
  title: 'Components/Utilities/Layout',
  component: Layout,
  parameters: { layout: 'fulscreen' },
} satisfies Meta<typeof Layout>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  // Just an empty `<div>` with a grid layout -- expected behaviour
  render: () => <Layout />,
};

export const ThreeColumns: Story = {
  name: 'Three column layout',
  render: () => {
    const [theme] = useTheme();

    const sectionStyle = css`
      grid-column: span 4;
      height: 100%;
      background-color: ${theme.color.display.blueMuted};
    `;

    const numberMarkerStyle = css`
      display: flex;
      justify-content: center;
      align-items: center;
      height: 8rem;
      font-size: ${theme.font.size.f20};
      font-weight: ${theme.font.weight.semiBold};
      color: ${theme.color.text.primary};
      font-family: ${theme.font.family.fixed};
      font-size: ${theme.font.size.f20};
    `;

    return (
      <Layout>
        <section className={sectionStyle}>
          <div className={numberMarkerStyle}>#1</div>
        </section>
        <section className={sectionStyle}>
          <div className={numberMarkerStyle}>#2</div>
        </section>
        <section className={sectionStyle}>
          <div className={numberMarkerStyle}>#3</div>
        </section>
      </Layout>
    );
  },
};

export const FourColumns: Story = {
  name: 'Four column layout',
  render: () => {
    const [theme] = useTheme();

    const sectionStyle = css`
      grid-column: span 4;
      height: 100%;
      background-color: ${theme.color.display.pinkMuted};

      ${theme.mq.mobile} {
        grid-column: span 2;
      }

      ${theme.mq.desktop} {
        grid-column: span 3;
      }
    `;

    const numberMarkerStyle = css`
      display: flex;
      justify-content: center;
      align-items: center;
      height: 8rem;
      font-size: ${theme.font.size.f20};
      font-weight: ${theme.font.weight.semiBold};
      color: ${theme.color.text.primary};
      font-family: ${theme.font.family.fixed};
      font-size: ${theme.font.size.f20};
    `;

    return (
      <Layout>
        <section className={sectionStyle}>
          <div className={numberMarkerStyle}>#1</div>
        </section>
        <section className={sectionStyle}>
          <div className={numberMarkerStyle}>#2</div>
        </section>
        <section className={sectionStyle}>
          <div className={numberMarkerStyle}>#3</div>
        </section>
        <section className={sectionStyle}>
          <div className={numberMarkerStyle}>#4</div>
        </section>
      </Layout>
    );
  },
};

export const GenericScreen: Story = {
  name: 'Generic screen',
  render: () => {
    const [theme] = useTheme();

    const mainHeaderStyle = css`
      grid-column: 1 / -1;
    `;

    const firstSectionStyle = css`
      grid-column: span 4;

      ${theme.mq.mobile} {
        grid-column: span 3;
      }
    `;

    const buttonGroupStyle = css`
      display: flex;
      flex-direction: column;
      gap: 16px;
    `;

    const secondSectionStyle = css`
      grid-column: span 4;

      ${theme.mq.tablet} {
        grid-column: span 5;
      }

      ${theme.mq.desktop} {
        grid-column: 5 / -2;
      }
    `;

    const thirdSectionStyle = css`
      grid-column: 1 / -1;

      ${theme.mq.desktop} {
        grid-column: 2 / -2;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 3 / -3;
      }
    `;

    const colourBlockStyle = css`
      background-color: ${theme.color.display.yellowVibrant};
      height: 250px;
    `;

    return (
      <Layout>
        <header className={mainHeaderStyle}>
          <Heading>UCL App</Heading>
        </header>

        <section className={firstSectionStyle}>
          <Heading level={2}>Some buttons</Heading>
          <div className={buttonGroupStyle}>
            <Button>Button 1</Button>
            <Button variant='secondary'>Button 2</Button>
          </div>
        </section>

        <section className={secondSectionStyle}>
          <Heading level={2}>Our story</Heading>
          <Text>
            UCL was founded on 11 February 1826 as an alternative to the
            Anglican universities of Oxford and Cambridge. It took the form of a
            joint stock company, with shares sold for £100 (equivalent to
            £10,600 in 2023) to proprietors, under the name of London
            University, although without legal recognition as a university or
            the associated right to award degrees. London University's first
            warden was Leonard Horner, who was the first scientist to head a
            British university.
          </Text>
        </section>

        <section className={thirdSectionStyle}>
          <Heading level={2}>A big wide block of colour </Heading>
          <div className={colourBlockStyle} />
        </section>
      </Layout>
    );
  },
};

export const CreativeGrid: Story = {
  name: 'Creative grid layout',
  render: () => {
    const [theme] = useTheme();

    const gridContainerStyle = css`
      grid-column: 1 / -1;
      display: grid;
      grid-template-columns: 1fr;
      gap: 16px;
      margin-bottom: 48px;

      ${theme.mq.mobile} {
        grid-template-columns: repeat(12, 1fr);
        grid-template-rows: repeat(6, minmax(80px, auto));
      }

      ${theme.mq.desktopLarge} {
        grid-template-columns: repeat(16, 1fr);
        grid-template-rows: repeat(8, minmax(90px, auto));
        gap: 24px;
      }
    `;

    const gridItemBaseStyle = css`
      display: flex;
      justify-content: center;
      align-items: center;
      font-weight: ${theme.font.weight.bold};
      color: white;
      border-radius: 4px;
      box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
      height: 120px;
      min-height: 100px;
      font-family: ${theme.font.family.fixed};
      font-size: ${theme.font.size.f20};
    `;

    const block1Style = css`
      ${gridItemBaseStyle}
      background-color: ${theme.color.display.purpleVibrant};

      ${theme.mq.mobile} {
        grid-column: 1 / 8;
        grid-row: 1 / 3;
        height: auto;
      }

      ${theme.mq.tablet} {
        grid-column: 1 / 7;
        grid-row: 1 / 4;
      }

      ${theme.mq.desktop} {
        grid-column: 1 / 6;
        grid-row: 1 / 3;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 2 / 6;
        grid-row: 2 / 4;
        border-radius: 50%;
        transform: rotate(-5deg);
      }
    `;

    const block2Style = css`
      ${gridItemBaseStyle}
      background-color: ${theme.color.display.blueVibrant};

      ${theme.mq.mobile} {
        grid-column: 8 / 13;
        grid-row: 1;
        height: auto;
      }

      ${theme.mq.tablet} {
        grid-column: 7 / 13;
        grid-row: 1 / 2;
      }

      ${theme.mq.desktop} {
        grid-column: 6 / 10;
        grid-row: 1;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 7 / 11;
        grid-row: 1 / 3;
        border-radius: 8px 8px 50% 50%;
      }
    `;

    const block3Style = css`
      ${gridItemBaseStyle}
      background-color: ${theme.color.display.greenVibrant};

      ${theme.mq.mobile} {
        grid-column: 8 / 13;
        grid-row: 2 / 4;
        height: auto;
      }

      ${theme.mq.tablet} {
        grid-column: 7 / 10;
        grid-row: 2 / 4;
      }

      ${theme.mq.desktop} {
        grid-column: 10 / 13;
        grid-row: 1 / 4;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 12 / 16;
        grid-row: 2 / 5;
        border-radius: 16px;
        transform: rotate(3deg);
      }
    `;

    const block4Style = css`
      ${gridItemBaseStyle}
      background-color: ${theme.color.display.blueMuted};

      ${theme.mq.mobile} {
        grid-column: 1 / 8;
        grid-row: 3 / 5;
        height: auto;
      }

      ${theme.mq.tablet} {
        grid-column: 10 / 13;
        grid-row: 2 / 5;
      }

      ${theme.mq.desktop} {
        grid-column: 6 / 10;
        grid-row: 2 / 4;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 7 / 11;
        grid-row: 4 / 6;
        clip-path: polygon(0% 0%, 100% 20%, 100% 100%, 0% 80%);
      }
    `;

    const block5Style = css`
      ${gridItemBaseStyle}
      background-color: ${theme.color.display.yellowVibrant};

      ${theme.mq.mobile} {
        grid-column: 1 / 8;
        grid-row: 5 / 7;
        height: auto;
      }

      ${theme.mq.tablet} {
        grid-column: 1 / 7;
        grid-row: 4 / 6;
      }

      ${theme.mq.desktop} {
        grid-column: 1 / 6;
        grid-row: 3 / 5;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 3 / 6;
        grid-row: 5 / 8;
        border-radius: 30% 70% 70% 30% / 30% 30% 70% 70%;
      }
    `;

    const block6Style = css`
      ${gridItemBaseStyle}
      background-color: ${theme.color.display.pinkVibrant};

      ${theme.mq.mobile} {
        grid-column: 8 / 13;
        grid-row: 4 / 7;
        height: auto;
      }

      ${theme.mq.tablet} {
        grid-column: 7 / 13;
        grid-row: 5 / 7;
      }

      ${theme.mq.desktop} {
        grid-column: 6 / 13;
        grid-row: 4 / 6;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 12 / 15;
        grid-row: 6 / 9;
        clip-path: polygon(50% 0%, 100% 50%, 50% 100%, 0% 50%);
      }
    `;

    const block7Style = css`
      ${gridItemBaseStyle}
      background-color: ${theme.color.display.blueVibrant};

      ${theme.mq.mobile} {
        grid-column: 1 / 13;
        grid-row: 7;
        height: auto;
      }

      ${theme.mq.tablet} {
        grid-column: 1 / 13;
        grid-row: 7 / 8;
      }

      ${theme.mq.desktop} {
        grid-column: 1 / 6;
        grid-row: 5 / 7;
      }

      ${theme.mq.desktopLarge} {
        grid-column: 7 / 10;
        grid-row: 7 / 9;
        border-radius: 0 50% 50% 50%;
        transform: rotate(10deg);
      }
    `;

    const headerStyle = css`
      grid-column: 1 / -1;
      text-align: center;
      margin-bottom: 32px;

      ${theme.mq.desktopLarge} {
        margin-bottom: 48px;
      }
    `;

    return (
      <Layout>
        <header className={headerStyle}>
          <Heading>Dynamic Grid Layout</Heading>
          <Text>
            CSS Grid can be used for essentially any placement of elements you
            could ever need..
          </Text>
        </header>

        <div className={gridContainerStyle}>
          <div className={block1Style}>#1</div>
          <div className={block2Style}>#2</div>
          <div className={block3Style}>#3</div>
          <div className={block4Style}>#4</div>
          <div className={block5Style}>#5</div>
          <div className={block6Style}>#6</div>
          <div className={block7Style}>#7</div>
        </div>
      </Layout>
    );
  },
};
