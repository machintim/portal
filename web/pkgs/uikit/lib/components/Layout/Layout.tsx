import { HTMLAttributes } from 'react';
import { css, cx } from '@emotion/css';
import { useTheme } from '../../theme';

export const NAME = 'ucl-uikit-layout';

export interface LayoutProps extends HTMLAttributes<HTMLDivElement> {
  testId?: string;
}

const Layout = ({
  children,
  testId = NAME,
  className,
  ...props
}: LayoutProps) => {
  const [theme] = useTheme();

  /**
   * Implements the Design System grid layout, as defined in the UCL Design System UI Kit
   * @see https://www.figma.com/design/8Sm5PxWOWJYpYXAzhRzUzt/UCL-Design-System-UI-Kit?node-id=1-286&p=f&t=bVPuyuMQfb0t2GTL-0
   *
   * CSS Implementation Details:
   * - CSS Grid layout with responsive columns: 4 (mobile), 8 (tablet), 12 (desktop)
   * - Max-width: 1536px
   *
   * - The Figma design uses the terms "Gutters" and "Margins":
   *   - "Gutters": Implemented using the `gap` property
   *   - "Margins": Implemented using the `padding` property, in order that `margin: 0 auto` centres the layout
   *
   * Responsive approach:
   * - Mobile-first: Default styles apply to all viewport widths (0px+)
   * - Media queries in `mq` blocks apply at specified breakpoints and wider
   */
  const baseStyle = css`
    display: grid;
    grid-template-columns: repeat(4, 1fr);
    gap: 16px; // "Gutter"
    max-width: ${theme.breakpoints.desktopLarge}px; // 1536px
    margin: 0 auto;
    padding: 0 16px; // "Margin"

    ${theme.mq.tablet} {
      grid-template-columns: repeat(8, 1fr);
      padding: 0 32px; // "Margin"
      gap: 24px; // "Gutter"
    }

    ${theme.mq.desktop} {
      grid-template-columns: repeat(12, 1fr);
      padding: 0 64px; // "Margin"
    }

    ${theme.mq.desktopLarge} {
      gap: 32px; // "Gutter"
    }

    // To prevent the content edges hitting the viewport boundary at desktopLarge breakpoint,
    // the breakpoint here is slightly wider.
    ${theme.mq.custom(1660)} {
      padding: 0; // "Margin" is removed, so the layout <div> takes up the full 1536px width
    }
  `;

  const style = cx(baseStyle, className);

  return (
    <div
      className={style}
      data-testid={testId}
      {...props}
    >
      {children}
    </div>
  );
};

export default Layout;
