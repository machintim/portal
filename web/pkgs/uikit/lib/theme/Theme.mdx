import { Meta, Title, Subtitle, Source } from '@storybook/addon-docs/blocks';
import theme from './defaultTheme';

{ console.log('theme:', theme) }

<Meta title="theme/Theme" />
<Title>Theme explorer</Title>
<Subtitle>All design tokens currently available in the theme</Subtitle>

This theme explorer lists all the named properties of the theme object, found in the <a href="https://github.com/ucl-isd/uikit-react/blob/main/lib/theme/defaultTheme.ts" target="_blank">`defaultTheme.ts` file</a> .

The theme object is a collection of <a href="https://atlassian.design/tokens/design-tokens/" target="_blank">design tokens</a> that are used to style the UIKit-React components.

The theme object is also logged to the console when this page is loaded. 

{
  Object.entries(theme).map(([key, value]) => {
  return (
    <div key={key}>
      <h2><code>{key}</code></h2>
      <Source language="json" code={JSON.stringify(value, null, 2)} />
    </div>
  );
})
}
