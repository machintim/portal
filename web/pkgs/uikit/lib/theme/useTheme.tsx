// We will probably move away from context/provider to store/hook for accessing theme.

import {
  createContext,
  useState,
  useContext,
  ReactNode,
} from 'react';

import theme, { ThemeType } from './defaultTheme';

const useThemeState = () => useState(theme as ThemeType);

const ThemeContext = createContext<ReturnType<
  typeof useThemeState
> | null>(null);

export const ThemeContextProvider: React.FC<{
  children: ReactNode;
}> = ({ children }) => (
  <ThemeContext.Provider value={useThemeState()}>
    {children}
  </ThemeContext.Provider>
);

const useTheme = () => {
  const value = useContext(ThemeContext);
  if (value === null) {
    throw new Error('Please add ThemeContextProvider');
  }
  return value;
};

export default useTheme;
