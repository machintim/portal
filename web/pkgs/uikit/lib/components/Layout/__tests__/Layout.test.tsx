import { describe, expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import { ThemeContextProvider } from '../../../theme/useTheme';
import Layout from '../Layout';

describe('Layout', () => {
  // Snapshot tests

  test('Snapshot: no children', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Layout />
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  test('Snapshot: with children', () => {
    const renderResult = render(
      <ThemeContextProvider>
        <Layout>
          <div>Child 1</div>
          <div>Child 2</div>
          <div>Child 3</div>
        </Layout>
      </ThemeContextProvider>
    );
    expect(renderResult.container.firstChild).toMatchSnapshot();
  });

  // Behaviour tests

  test('Can be found by default testId', () => {
    const defaultTestId = 'ucl-uikit-layout';
    render(
      <ThemeContextProvider>
        <Layout />
      </ThemeContextProvider>
    );
    expect(screen.getByTestId(defaultTestId)).toBeInTheDocument();
  });

  test('Can be found by testId', () => {
    render(
      <ThemeContextProvider>
        <Layout testId='test-id' />
      </ThemeContextProvider>
    );
    expect(screen.getByTestId('test-id')).toBeInTheDocument();
  });
});
