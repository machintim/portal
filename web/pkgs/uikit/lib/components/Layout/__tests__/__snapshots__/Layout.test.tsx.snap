// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Layout > Snapshot: no children 1`] = `
<div
  class="css-3reham"
  data-testid="ucl-uikit-layout"
/>
`;

exports[`Layout > Snapshot: with children 1`] = `
<div
  class="css-3reham"
  data-testid="ucl-uikit-layout"
>
  <div>
    Child 1
  </div>
  <div>
    Child 2
  </div>
  <div>
    Child 3
  </div>
</div>
`;
